### DISCUSSION TITLE
Expect a particular response

### DISCUSSION THREAD

**Discussion by Lex-Kent** [speechly/speechly#238] [Q&A] [answered]
Hi all, I'm only just getting to grips with Speechly and I have what is probably a dumb question!

I have a virtual assistant that chats with my user to help them with certain aspects of a complex financial tool. There are lots of questions, but the one I'm struggling with most is when the response is a single letter. My assistant reads (and shows on screen) 4 options for a particular input and asks

"Would you consider yourself more like A, B, C or D?"

To which the user responds (generally) with a single letter. however, Speechly (and webkitSpeechRecognition) hears things like "day", "sea", "say", "eh". I can work around these, but it would be far easier if I could tell Speechly to expect one of "A", "B", "C" or "D" before anything else. Can this be done? I think it's called "grammar" with some speech recognition APIs, but I'm not entirely sure - like I said, this is pretty new to me!

Thanks in advance for any pointers...

**Comment by bigdatabaracus** [comment]
Hello @Lex-Kent, yes, you can define a "grammar" with Speechly to improve the recognition accuracy but to also format Speechly's response in a way that is easier for you to use down stream. 

In the Speechly lingo grammars are called "configurations". For more information on how to define your configuration, you can check out the [intents and entities section](https://docs.speechly.com/features/intents-entities) our docs.  

However to give you an idea your configuration will probably need to look something like this 

```
*response [A](alternative)
*response [B](alternative)
*response [C](alternative)
*response [D](alternative)
```

Hope this helped! 

best regards, Hannes

**Reply by Lex-Kent** [reply]
Thanks Hannes,

I really appreciate you taking the time to help.

I only want this particular restriction to apply to this question, though. Earlier in the tool, my assistant asks the user for their name. I wouldn't want a user called "Dee" to have her response reduced to "D". Can I target a particular part of the configuration with each question?

**Reply by bigdatabaracus** [reply]
Hello @Lex-Kent Depending on the state your application is in you can call different Speechly applications (app id). So if your application is expecting the "A", "B", "C" or "D" answer, you can call speechly with of the app id with a corresponding configuration. If you are expecting something else you can use a different app id, with a different configuration or potentially even an empty configuration. 

Does this make sense?

**Reply by Lex-Kent** [reply]
Thanks Hannes, yes, that makes sense. It's a shame though - I'd have reconfigure my entire application. At the moment I have a single speechly instance and I simply stop and start it when my assistant speaks or asks a question.

I would be great if I could pass the grammar with the `.start()` event of my existing instance, but never mind.

For now I think I'll just have an array of possible responses including "day", "see", "eh" etc. and match each of those to a letter so I can work with my single Speechly instance.

Thanks for your help, it's much appreciated!

**Reply by Lex-Kent** [reply]
Hi @bigdatabaracus . I've been working on other aspects of my tool, but I've circled back to this. I've added an application for this question/answer as you suggested. My training data is very simple, it looks like this:

```
*chooseRisk [A](riskLetter)
*chooseRisk [B](riskLetter)
*chooseRisk [C](riskLetter)
*chooseRisk [D](riskLetter)
```

but Speechly still hears (for example) "Say" when I reply "C". Is there anything else I need to do to prevent this from happening please?

Many thanks again

**Reply by bigdatabaracus** [reply]
hello @Lex-Kent please check out https://docs.speechly.com/features/imports-lookups#lookups. It's a hack, but you can define a mapping where "say" maps to "C".

**Reply by Lex-Kent** [reply]
Ooh, thanks @bigdatabaracus I'll take a look...

**Reply by Lex-Kent** [reply]
I've had a go at creating a lookup csv and I've uploaded it to my application dashboard, but I can no longer deploy the app. Having imported my csv, I see a new lookup table:

![image](https://user-images.githubusercontent.com/107930534/220088904-4efe9383-e1e2-4228-848c-5065a6b9b352.png)

and I changed my Text adapation to

`*chooseRisk $abcd_lookups(riskLetter)`

but my deployment now fails. I don't know why it fails - there is no error log anywhere that I can see.

What am I doing wrong please @bigdatabaracus !?

**Reply by bigdatabaracus** [reply]
@teelisyys could you be so kind and help @Lex-Kent out here? Thanks!

**Reply by Lex-Kent** [reply]
**Update:** @bigdatabaracus and @teelisyys. I deleted the `riskLetter` entity, deleted the csv lookup and created a new lookup called `letters` from scratch in the dashboard. My Text adaptation now looks like

`*chooseRisk $letters(letters)`

and it now deploys. However, I'm not actually getting lookup replacements in my responses. Having this lookup data:

![image](https://user-images.githubusercontent.com/107930534/220113996-26cbb403-fe44-4924-b1f3-44a6b701f6d5.png)

gives me these responses:

![image](https://user-images.githubusercontent.com/107930534/220114499-a88c7adc-c532-4367-ac5d-dc7eaf1ef654.png)

and switching the columns in the lookup table doesn't seem to make any difference!

**Comment by teelisyys** [comment, marked as answer]
Hi @Lex-Kent !

Please note that on the _Preview_ page, the _output_ value, called _entity value_ here, is show only when hovering the entity. The value in the entity box is always the _input_ value. Could you verify if there is a hover like this?

<img width="212" alt="image" src="https://user-images.githubusercontent.com/9036453/220122355-6c4993fa-c741-49e3-90be-5ef4cb4785ed.png">


The latest config you posted looks indeed correct to me. Only potential issue is the having multiple input values that only differ in casing. This can lead to unexpected results with that input, but other inputs should not be affected.

As you noted the latest deployment was successful. The older deployments were failing due to an empty item in the lookup csv. This should have also have been reported as the cause of the failure, and it is not immediately obvious why that was not the case.

**Reply by Lex-Kent** [reply]
Thanks @teelisyys I am indeed seeing the entity value on hover. So I guess it's all working!

On a side note, I noticed that the previous deployment time is wrong when I redeploy my training set:

![image](https://user-images.githubusercontent.com/107930534/220127496-4285e3d5-38f3-4eed-8ad4-4c963f32d145.png)

on the screen grab above, the previous deployment time should have been somewhere around 3½ minutes.

Thanks again both :o)
